src/datachain/catalog/catalog.py

```diff
@@ -509,6 +509,60 @@ def clone_catalog_with_cache(catalog: "Catalog", cache: "Cache") -> "Catalog":
     return clone
 
 
+def get_all_ids(d: dict[str, Any]) -> list[int]:
+    """
+    Extracts all unique dataset IDs from a nested dictionary structure.
+    """
+    ids = set()
+    for key, value in d.items():
+        ids.add(int(key))
+        if isinstance(value, dict):
+            ids.update(get_all_ids(value))
+    return sorted(ids)
+
+
+def build_nested_dependencies(
+    dataset_deps: dict[str, DatasetDependency],
+    dependency_structure: dict[str, dict],
+    indirect: bool = True,
```

Comment:
do we use it in both modes? why?

Reply:
This is for convenience, it fetches one level and return if indirect is false. I will add a test to verify this further and make changes if needed.